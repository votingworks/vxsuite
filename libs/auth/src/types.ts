/**
 * Valid values for cert card type field
 */
export type CardType =
  | 'system-administrator'
  | 'election-manager'
  | 'poll-worker'
  | 'poll-worker-with-pin';

/**
 * Parsed custom cert fields
 */
export interface CustomCertFields {
  component: 'card' | 'admin' | 'central-scan' | 'mark' | 'scan';
  jurisdiction: string;
  cardType?: CardType;
  electionHash?: string;
}
