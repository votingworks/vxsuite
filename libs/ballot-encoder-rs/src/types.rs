use std::num::NonZeroU8;

use static_assertions::const_assert;

use crate::codable;

codable!(PrecinctIndex, u32, 0..=4096);
codable!(BallotStyleIndex, u32, 0..=4096);
codable!(PageNumber, u8, 1..=30);

// Statically validate our maximum values fit within the types we're using.
// TODO: move this into `codable!`
const_assert!(PrecinctIndex::BITS <= usize::BITS);
const_assert!(BallotStyleIndex::BITS <= usize::BITS);
const_assert!(PageNumber::BITS <= u8::BITS);

impl PageNumber {
    /// Whether this is the first page of its sheet, i.e. the front.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = PageNumber::new_unchecked(2);
    ///
    /// assert!(page_one.is_recto());
    /// assert!(!page_two.is_recto());
    /// ```
    #[must_use]
    pub const fn is_recto(self) -> bool {
        self.0 % 2 == 1
    }

    /// Whether this is the second page of its sheet, i.e. the back.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = PageNumber::new_unchecked(2);
    ///
    /// assert!(!page_one.is_verso());
    /// assert!(page_two.is_verso());
    /// ```
    #[must_use]
    pub const fn is_verso(self) -> bool {
        self.0 % 2 == 0
    }

    /// Determines the sheet number.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = PageNumber::new_unchecked(2);
    /// let page_three = PageNumber::new_unchecked(3);
    /// let page_four = PageNumber::new_unchecked(4);
    ///
    /// assert_eq!(page_one.sheet_number().get(), 1);
    /// assert_eq!(page_two.sheet_number().get(), 1);
    /// assert_eq!(page_three.sheet_number().get(), 2);
    /// assert_eq!(page_four.sheet_number().get(), 2);
    /// ```
    ///
    /// # Panics
    ///
    /// If `Self` is an invalid `PageNumber` that would produce a sheet number
    /// of zero, this method will panic.
    #[must_use]
    pub const fn sheet_number(self) -> NonZeroU8 {
        let verso_page = if self.is_verso() {
            self
        } else {
            self.opposite()
        };
        NonZeroU8::new(verso_page.get() / 2).expect("sheet number >= 1")
    }

    /// Gets the `PageNumber` opposite this one on the same sheet.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = page_one.opposite();
    /// assert_eq!(page_two.get(), 2);
    /// ```
    pub const fn opposite(self) -> Self {
        match self.0 % 2 {
            0 => Self(self.0 - 1),
            1 => Self(self.0 + 1),
            _ => unreachable!(),
        }
    }
}

#[cfg(test)]
pub mod tests {
    use std::num::NonZeroU8;

    use proptest::{prelude::Strategy, proptest};

    use super::PageNumber;

    #[test]
    fn test_page_number() {
        let page_one = PageNumber::new_unchecked(1);
        let page_two = PageNumber::new_unchecked(2);

        assert_eq!(page_one.get(), 1);
        assert_eq!(page_two.get(), 2);

        assert_eq!(page_one, page_one);
        assert_ne!(page_one, page_two);

        assert!(page_one.is_recto());
        assert!(!page_two.is_recto());
        assert!(!page_one.is_verso());
        assert!(page_two.is_verso());

        assert_eq!(page_one.sheet_number(), NonZeroU8::new(1).unwrap());
        assert_eq!(page_two.sheet_number(), NonZeroU8::new(1).unwrap());
    }

    pub fn arbitrary_page_number() -> impl Strategy<Value = PageNumber> {
        (PageNumber::MIN_VALUE..=PageNumber::MAX_VALUE).prop_map(PageNumber::new_unchecked)
    }

    proptest! {
        #[test]
        fn test_valid_page_number(page_number in arbitrary_page_number()) {
            assert_eq!(page_number, page_number.opposite().opposite());
            assert_eq!(page_number.is_recto(), page_number.opposite().is_verso());
            assert_eq!(page_number.is_verso(), page_number.opposite().is_recto());
            assert_eq!(page_number.sheet_number(), page_number.opposite().sheet_number());
        }

        #[test]
        fn test_invalid_page_number(value in PageNumber::MAX_VALUE + 1..=u8::MAX) {
            assert_eq!(PageNumber::new(value), None);
        }
    }
}
