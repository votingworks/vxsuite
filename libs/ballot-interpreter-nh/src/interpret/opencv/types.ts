import * as cv from '@u4/opencv4nodejs';

import {
  GridLayout,
  GridPosition,
  HmpbBallotPageMetadata,
  PageInterpretation,
  Result,
} from '@votingworks/types';

/** TODO */
export interface PageImageData {
  binarizedImage: cv.Mat;
  optimizedImage: cv.Mat;
  originalImage: cv.Mat;
}

/** Single page interpretation result. */
export interface OpenCvInterpretResult {
  interpretation: PageInterpretation;
  normalizedImageFilePath: string;
  originalImageFilePath: string;
}

/** TODO */
export interface PageRegion {
  image: cv.Mat;
  offset: cv.Point2;
}

/** TODO */
export interface PageMargins<T> {
  bottom: T;
  left: T;
  right: T;
  top: T;
}

/** TODO */
export type MarginType = keyof PageMargins<unknown>;

/** TODO */
export interface PotentialTimingMarks {
  allContours: cv.Contour[];
  potentialMarks: cv.Contour[];
}

/** TODO */
export interface DetectedTimingMarks {
  marginImage: PageRegion;
  marginType: MarginType;
  marks: cv.Contour[];
  medianHeight: number;
  medianSpacing: number;
  medianWidth: number;
  // pageRegion: PageRegion;
}

/** TODO */
export interface TimingMarks {
  detected: cv.Contour[];
  first: cv.Rect;
  interpolated: cv.Rect[];
  last: cv.Rect;
  medianSpacing: number;
  marginType: MarginType;
}

/** TODO */
export type BallotMetadata = Pick<
  HmpbBallotPageMetadata,
  'ballotStyleId' | 'ballotType' | 'electionHash' | 'isTestMode' | 'precinctId'
> & {
  gridLayout: GridLayout;
};

/** TODO */
export interface OvalTemplate {
  image: cv.Mat;
  shadedPixelRatio: number;
}

/** TODO */
export interface OvalMark {
  readonly gridPosition: GridPosition;
  readonly expectedBounds: Result<cv.Rect, string>;
  readonly matchBounds: Result<cv.Rect, string>;
  readonly score: number;
  readonly searchBounds: Result<cv.Rect, string>;
}
