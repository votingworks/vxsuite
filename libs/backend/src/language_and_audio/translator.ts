import { TranslationServiceClient as GoogleCloudTranslationClient } from '@google-cloud/translate';
import { assertDefined, iter } from '@votingworks/basics';

import { NonEnglishLanguageCode, LanguageCode } from '@votingworks/types';
import { GOOGLE_CLOUD_PROJECT_ID } from './google_cloud_config';

/**
 * The subset of {@link GoogleCloudTranslationClient} that we actually use
 */
export type MinimalGoogleCloudTranslationClient = Pick<
  GoogleCloudTranslationClient,
  'translateText'
>;

/**
 * Interface for a translator that can translate text to a specified language.
 */
export interface Translator {
  translateText(
    textArray: string[],
    targetLanguageCode: NonEnglishLanguageCode
  ): Promise<string[]>;
}

/**
 * A simple base class that provides a utility function to translate text with google cloud.
 * When used directly this translator implementation will not cache translations.
 * Caching or other app-specific needs should be handled in a specific sub-class implementation.
 */
export class GoogleCloudTranslator implements Translator {
  private readonly translationClient: MinimalGoogleCloudTranslationClient;

  constructor(input: {
    // Support providing a mock client for tests
    translationClient?: MinimalGoogleCloudTranslationClient;
  }) {
    this.translationClient =
      input.translationClient ??
      /* istanbul ignore next */ new GoogleCloudTranslationClient();
  }

  async translateText(
    textArray: string[],
    targetLanguageCode: NonEnglishLanguageCode
  ): Promise<string[]> {
    return await this.translateTextWithGoogleCloud(
      textArray,
      targetLanguageCode
    );
  }

  protected async translateTextWithGoogleCloud(
    textArray: string[],
    targetLanguageCode: NonEnglishLanguageCode
  ): Promise<string[]> {
    // Google Cloud will preserve HTML tags fairly well, so we can pass HTML
    // rich text directly to the API. However, it has a max string length limit,
    // so base64 encoded img src attributes are generally too long to include.
    // We strip them out in order and replace them after translating.
    const srcRegex = /src="([^"]*)"/g;
    const srcAttrsArray = textArray.map((text) =>
      iter(text.matchAll(srcRegex))
        .map((match) => match[0])
        .toArray()
    );

    function srcPlaceholder(index: number) {
      return `src="${index}"`;
    }
    const textArrayWithoutSrcAttrs = iter(textArray)
      .zip(srcAttrsArray)
      .map(([textString, srcAttrs]) =>
        srcAttrs.reduce(
          (text, src, i) => text.replace(src, srcPlaceholder(i)),
          textString
        )
      )
      .toArray();

    const [response] = await this.translationClient.translateText({
      contents: textArrayWithoutSrcAttrs,
      mimeType: 'text/plain',
      parent: `projects/${GOOGLE_CLOUD_PROJECT_ID}`,
      sourceLanguageCode: LanguageCode.ENGLISH,
      targetLanguageCode,
    });

    const translatedTextArrayWithSrcAttrs = iter(response.translations)
      .zip(srcAttrsArray)
      .map(([{ translatedText }, srcAttrs]) =>
        srcAttrs.reduce(
          (text, src, i) => text.replace(srcPlaceholder(i), src),
          assertDefined(translatedText)
        )
      )
      .toArray();

    return translatedTextArrayWithSrcAttrs;
  }
}
