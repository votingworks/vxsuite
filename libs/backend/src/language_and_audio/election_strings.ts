import { assert, assertDefined } from '@votingworks/basics';
import {
  CandidateContest,
  Election,
  ElectionStringKey,
  UiStringsPackage,
  YesNoContest,
  BallotLanguageConfigs,
  getAllBallotLanguages,
  LanguageCode,
} from '@votingworks/types';

import { format } from '@votingworks/utils';
import { GoogleCloudTranslator } from './translator';
import { setUiString } from './utils';

interface ElectionString {
  stringKey: ElectionStringKey | [ElectionStringKey, string];
  stringInEnglish: string;
}

interface ElectionStringConfigNotTranslatable {
  translatable: false;
}

interface ElectionStringConfigTranslatable {
  translatable: true;
  customTranslationMethod?: (input: {
    election: Election;
    languageCode: LanguageCode;
    stringInEnglish: string;
  }) => string;
}

type ElectionStringConfig =
  | ElectionStringConfigNotTranslatable
  | ElectionStringConfigTranslatable;

const electionStringConfigs: Record<ElectionStringKey, ElectionStringConfig> = {
  [ElectionStringKey.BALLOT_LANGUAGE]: {
    translatable: true,
    customTranslationMethod: (p) =>
      format.languageDisplayName({
        languageCode: p.languageCode,
      }),
  },
  [ElectionStringKey.BALLOT_STYLE_ID]: {
    translatable: false,
  },
  [ElectionStringKey.CANDIDATE_NAME]: {
    translatable: false,
  },
  [ElectionStringKey.CONTEST_DESCRIPTION]: {
    translatable: true,
  },
  [ElectionStringKey.CONTEST_OPTION_LABEL]: {
    translatable: true,
  },
  [ElectionStringKey.CONTEST_TERM]: {
    translatable: true,
  },
  [ElectionStringKey.CONTEST_TITLE]: {
    translatable: true,
  },
  [ElectionStringKey.COUNTY_NAME]: {
    translatable: true,
  },
  [ElectionStringKey.DISTRICT_NAME]: {
    translatable: true,
  },
  [ElectionStringKey.ELECTION_DATE]: {
    translatable: true,
    customTranslationMethod: ({ election, languageCode }) =>
      format.localeLongDate(
        election.date.toMidnightDatetimeWithSystemTimezone(),
        languageCode
      ),
  },
  [ElectionStringKey.ELECTION_TITLE]: {
    translatable: true,
  },
  [ElectionStringKey.PARTY_FULL_NAME]: {
    translatable: true,
  },
  [ElectionStringKey.PARTY_NAME]: {
    translatable: true,
  },
  [ElectionStringKey.PRECINCT_NAME]: {
    translatable: true,
  },
  [ElectionStringKey.STATE_NAME]: {
    translatable: true,
  },
};

const electionStringExtractorFns: Record<
  ElectionStringKey,
  (election: Election) => ElectionString[]
> = {
  [ElectionStringKey.BALLOT_LANGUAGE]() {
    return [
      {
        stringKey: ElectionStringKey.BALLOT_LANGUAGE,
        stringInEnglish: 'English',
      },
    ];
  },
  [ElectionStringKey.BALLOT_STYLE_ID](election) {
    return election.ballotStyles.map((ballotStyle) => ({
      stringKey: [ElectionStringKey.BALLOT_STYLE_ID, ballotStyle.id],
      stringInEnglish: ballotStyle.groupId,
    }));
  },
  [ElectionStringKey.CANDIDATE_NAME](election) {
    return election.contests
      .filter(
        (contest): contest is CandidateContest => contest.type === 'candidate'
      )
      .flatMap((contest) =>
        contest.candidates.map(
          (candidate): ElectionString => ({
            stringKey: [ElectionStringKey.CANDIDATE_NAME, candidate.id],
            stringInEnglish: candidate.name,
          })
        )
      );
  },
  [ElectionStringKey.CONTEST_DESCRIPTION](election) {
    return election.contests
      .filter((contest): contest is YesNoContest => contest.type === 'yesno')
      .map((contest) => ({
        stringKey: [ElectionStringKey.CONTEST_DESCRIPTION, contest.id],
        stringInEnglish: contest.description,
      }));
  },
  [ElectionStringKey.CONTEST_OPTION_LABEL](election) {
    return election.contests
      .filter((contest): contest is YesNoContest => contest.type === 'yesno')
      .flatMap((contest): ElectionString[] => [
        {
          stringKey: [
            ElectionStringKey.CONTEST_OPTION_LABEL,
            contest.yesOption.id,
          ],
          stringInEnglish: contest.yesOption.label,
        },
        {
          stringKey: [
            ElectionStringKey.CONTEST_OPTION_LABEL,
            contest.noOption.id,
          ],
          stringInEnglish: contest.noOption.label,
        },
      ]);
  },
  [ElectionStringKey.CONTEST_TERM](election) {
    return election.contests
      .filter(
        (contest): contest is CandidateContest => contest.type === 'candidate'
      )
      .filter((contest) => contest.termDescription)
      .map((contest) => ({
        stringKey: [ElectionStringKey.CONTEST_TERM, contest.id],
        stringInEnglish: assertDefined(contest.termDescription),
      }));
  },
  [ElectionStringKey.CONTEST_TITLE](election) {
    return election.contests.map((contest) => ({
      stringKey: [ElectionStringKey.CONTEST_TITLE, contest.id],
      stringInEnglish: contest.title,
    }));
  },
  [ElectionStringKey.COUNTY_NAME](election) {
    return [
      {
        stringKey: ElectionStringKey.COUNTY_NAME,
        stringInEnglish: election.county.name,
      },
    ];
  },
  [ElectionStringKey.DISTRICT_NAME](election) {
    return election.districts.map((district) => ({
      stringKey: [ElectionStringKey.DISTRICT_NAME, district.id],
      stringInEnglish: district.name,
    }));
  },
  [ElectionStringKey.ELECTION_DATE](election) {
    return [
      {
        stringKey: ElectionStringKey.ELECTION_DATE,
        stringInEnglish: format.localeLongDate(
          election.date.toMidnightDatetimeWithSystemTimezone(),
          LanguageCode.ENGLISH
        ),
      },
    ];
  },
  [ElectionStringKey.ELECTION_TITLE](election) {
    return [
      {
        stringKey: ElectionStringKey.ELECTION_TITLE,
        stringInEnglish: election.title,
      },
    ];
  },
  [ElectionStringKey.PARTY_FULL_NAME](election) {
    return election.parties.map((party) => ({
      stringKey: [ElectionStringKey.PARTY_FULL_NAME, party.id],
      stringInEnglish: party.fullName,
    }));
  },
  [ElectionStringKey.PARTY_NAME](election) {
    return election.parties.map((party) => ({
      stringKey: [ElectionStringKey.PARTY_NAME, party.id],
      stringInEnglish: party.name,
    }));
  },
  [ElectionStringKey.PRECINCT_NAME](election) {
    return election.precincts.map((precinct) => ({
      stringKey: [ElectionStringKey.PRECINCT_NAME, precinct.id],
      stringInEnglish: precinct.name,
    }));
  },
  [ElectionStringKey.STATE_NAME](election) {
    return [
      {
        stringKey: ElectionStringKey.STATE_NAME,
        stringInEnglish: election.state,
      },
    ];
  },
};

function getElectionStringConfig({
  stringKey,
}: ElectionString): ElectionStringConfig {
  return typeof stringKey === 'string'
    ? electionStringConfigs[stringKey]
    : electionStringConfigs[stringKey[0]];
}

function extractElectionStrings(election: Election): ElectionString[] {
  return Object.values(electionStringExtractorFns).flatMap((extractorFn) =>
    extractorFn(election)
  );
}

/**
 * Finds all election strings that need translation for the given election and set of ballot language configs and returns
 * the translated results.
 */
export async function extractAndTranslateElectionStrings(
  translator: GoogleCloudTranslator,
  election: Election,
  ballotLanguageConfigs: BallotLanguageConfigs
): Promise<UiStringsPackage> {
  const languages = getAllBallotLanguages(ballotLanguageConfigs);
  const untranslatedElectionStrings = extractElectionStrings(election);
  const electionStringsNotToTranslate = untranslatedElectionStrings.filter(
    (electionString) => {
      const config = getElectionStringConfig(electionString);
      return !config.translatable;
    }
  );
  const electionStringsToCloudTranslate = untranslatedElectionStrings.filter(
    (electionString) => {
      const config = getElectionStringConfig(electionString);
      return config.translatable && !config.customTranslationMethod;
    }
  );
  const electionStringsToCustomTranslate = untranslatedElectionStrings.filter(
    (electionString) => {
      const config = getElectionStringConfig(electionString);
      return config.translatable && config.customTranslationMethod;
    }
  );

  const electionStrings: UiStringsPackage = {};

  // Election strings not to translate
  for (const electionString of electionStringsNotToTranslate) {
    setUiString(
      electionStrings,
      LanguageCode.ENGLISH,
      electionString.stringKey,
      electionString.stringInEnglish
    );
  }

  // Election strings to cloud translate
  const stringsInEnglish = electionStringsToCloudTranslate.map(
    ({ stringInEnglish }) => stringInEnglish
  );
  for (const languageCode of languages) {
    const stringsInLanguage =
      languageCode === LanguageCode.ENGLISH
        ? stringsInEnglish
        : await translator.translateText(stringsInEnglish, languageCode);
    for (const [
      i,
      electionString,
    ] of electionStringsToCloudTranslate.entries()) {
      const { stringKey } = electionString;
      const stringInLanguage = stringsInLanguage[i];
      const config = getElectionStringConfig(electionString);
      assert(config.translatable);
      setUiString(
        electionStrings,
        languageCode,
        stringKey,
        assertDefined(stringInLanguage)
      );
    }
  }

  // Election strings to custom translate
  for (const electionString of electionStringsToCustomTranslate) {
    const { stringKey, stringInEnglish } = electionString;
    const config = getElectionStringConfig(electionString);
    assert(config.translatable && config.customTranslationMethod);
    for (const languageCode of languages) {
      const stringInLanguage = config.customTranslationMethod({
        election,
        languageCode,
        stringInEnglish,
      });
      setUiString(electionStrings, languageCode, stringKey, stringInLanguage);
    }
  }

  return electionStrings;
}
