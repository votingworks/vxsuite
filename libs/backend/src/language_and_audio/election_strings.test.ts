import {
  electionPrimaryPrecinctSplitsFixtures,
  electionTwoPartyPrimaryFixtures,
} from '@votingworks/fixtures';
import { LanguageCode, BallotLanguageConfigs } from '@votingworks/types';
import { assert } from '@votingworks/basics';
import { extractAndTranslateElectionStrings } from './election_strings';
import { GoogleCloudTranslator } from './translator';
import { MockGoogleCloudTranslationClient } from './test_utils';

const englishOnlyConfig: BallotLanguageConfigs = [
  { languages: [LanguageCode.ENGLISH] },
];
const englishSpanishLanguageConfig: BallotLanguageConfigs = [
  { languages: [LanguageCode.ENGLISH, LanguageCode.SPANISH] },
];

describe('extractAndTranslateElectionStrings', () => {
  it('should extract and translate election strings correctly for english only', async () => {
    const translationClient = new MockGoogleCloudTranslationClient();
    const mockTranslator = new GoogleCloudTranslator({ translationClient });
    const result = await extractAndTranslateElectionStrings(
      mockTranslator,
      electionPrimaryPrecinctSplitsFixtures.readElection(),
      englishOnlyConfig
    );

    expect(result).toBeDefined();
    expect(Object.keys(result)).toEqual([LanguageCode.ENGLISH]);
    const englishResults = result[LanguageCode.ENGLISH];
    assert(englishResults);
    expect(englishResults).toMatchSnapshot();
  });

  it('should extract and translate election strings correctly for english only with election with contest term', async () => {
    const translationClient = new MockGoogleCloudTranslationClient();
    const mockTranslator = new GoogleCloudTranslator({ translationClient });
    const result = await extractAndTranslateElectionStrings(
      mockTranslator,
      electionTwoPartyPrimaryFixtures.readElection(),
      englishOnlyConfig
    );

    expect(result).toBeDefined();
    expect(Object.keys(result)).toEqual([LanguageCode.ENGLISH]);
    const englishResults = result[LanguageCode.ENGLISH];
    assert(englishResults);
    expect(englishResults).toMatchSnapshot();
  });

  it('should extract and translate election strings correctly for multiple languages', async () => {
    const translationClient = new MockGoogleCloudTranslationClient();
    const mockTranslator = new GoogleCloudTranslator({ translationClient });
    const result = await extractAndTranslateElectionStrings(
      mockTranslator,
      electionPrimaryPrecinctSplitsFixtures.readElection(),
      englishSpanishLanguageConfig
    );

    expect(result).toBeDefined();
    expect(Object.keys(result)).toEqual([
      LanguageCode.ENGLISH,
      LanguageCode.SPANISH,
    ]);
    const englishResults = result[LanguageCode.ENGLISH];
    const spanishResults = result[LanguageCode.SPANISH];
    assert(englishResults);
    assert(spanishResults);
    // Should translate all the same fields
    expect(spanishResults).toMatchSnapshot();
  });
});
