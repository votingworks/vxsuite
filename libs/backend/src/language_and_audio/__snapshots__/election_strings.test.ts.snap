// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractAndTranslateElectionStrings should extract and translate election strings correctly for english only 1`] = `
{
  "ballotLanguage": "English",
  "ballotStyleId": {
    "1-F_en": "1-F",
    "1-F_es-US": "1-F",
    "1-F_zh-Hans": "1-F",
    "1-F_zh-Hant": "1-F",
    "1-Ma_en": "1-Ma",
    "1-Ma_es-US": "1-Ma",
    "1-Ma_zh-Hans": "1-Ma",
    "1-Ma_zh-Hant": "1-Ma",
    "2-F_en": "2-F",
    "2-F_es-US": "2-F",
    "2-F_zh-Hans": "2-F",
    "2-F_zh-Hant": "2-F",
    "2-Ma_en": "2-Ma",
    "2-Ma_es-US": "2-Ma",
    "2-Ma_zh-Hans": "2-Ma",
    "2-Ma_zh-Hant": "2-Ma",
    "3-F_en": "3-F",
    "3-F_es-US": "3-F",
    "3-F_zh-Hans": "3-F",
    "3-F_zh-Hant": "3-F",
    "3-Ma_en": "3-Ma",
    "3-Ma_es-US": "3-Ma",
    "3-Ma_zh-Hans": "3-Ma",
    "3-Ma_zh-Hant": "3-Ma",
    "4-F_en": "4-F",
    "4-F_es-US": "4-F",
    "4-F_zh-Hans": "4-F",
    "4-F_zh-Hant": "4-F",
    "4-Ma_en": "4-Ma",
    "4-Ma_es-US": "4-Ma",
    "4-Ma_zh-Hans": "4-Ma",
    "4-Ma_zh-Hant": "4-Ma",
  },
  "candidateName": {
    "elephant-1": "Elephant",
    "elephant-2": "Elephant",
    "fox": "Fox",
    "horse": "Horse",
    "kangaroo-1": "Kangaroo",
    "kangaroo-2": "Kangaroo",
    "lion-1": "Lion",
    "lion-2": "Lion",
    "manta-ray-1": "Manta Ray",
    "manta-ray-2": "Manta Ray",
    "otter": "Otter",
    "pufferfish-1": "Pufferfish",
    "pufferfish-2": "Pufferfish",
    "rockfish-1": "Rockfish",
    "rockfish-2": "Rockfish",
    "salmon": "Salmon",
    "seahorse": "Seahorse",
    "triggerfish-1": "Triggerfish",
    "triggerfish-2": "Triggerfish",
    "zebra-1": "Zebra",
    "zebra-2": "Zebra",
  },
  "contestDescription": {
    "water-1-fishing": "Should fishing be banned in all city owned lakes and rivers?",
    "water-2-fishing": "Should fishing be banned in all city owned lakes and rivers?",
  },
  "contestOptionLabel": {
    "water-1-fishing-allow-fishing": "NO",
    "water-1-fishing-ban-fishing": "YES",
    "water-2-fishing-allow-fishing": "NO",
    "water-2-fishing-ban-fishing": "YES",
  },
  "contestTitle": {
    "congressional-1-fish": "Congressional 1 Representative - Fish",
    "congressional-1-mammal": "Congressional 1 Representative - Mammal",
    "congressional-2-fish": "Congressional 2 Representative - Fish",
    "congressional-2-mammal": "Congressional 2 Representative - Mammal",
    "county-leader-fish": "County Leader Fish",
    "county-leader-mammal": "County Leader Mammal",
    "water-1-fishing": "Water District 1 Fishing",
    "water-2-fishing": "Water District 2 Fishing",
  },
  "countyName": "Sample County",
  "districtName": {
    "congressional-1": "Congressional District 1",
    "congressional-2": "Congressional District 2",
    "county": "County",
    "water-1": "Water District 1",
    "water-2": "Water District 2",
  },
  "electionDate": "September 8, 2021",
  "electionTitle": "Example Primary Election",
  "partyFullName": {
    "0": "Mammal Party",
    "1": "Fish Party",
  },
  "partyName": {
    "0": "Mammal",
    "1": "Fish",
  },
  "precinctName": {
    "precinct-c1-w1-1": "Precinct 1",
    "precinct-c1-w1-2": "Precinct 2",
    "precinct-c1-w2": "Precinct 3",
    "precinct-c2": "Precinct 4",
  },
  "stateName": "State of Sample",
}
`;

exports[`extractAndTranslateElectionStrings should extract and translate election strings correctly for english only with election with contest term 1`] = `
{
  "ballotLanguage": "English",
  "ballotStyleId": {
    "1M": "1M",
    "2F": "2F",
  },
  "candidateName": {
    "elephant": "Elephant",
    "fox": "Fox",
    "horse": "Horse",
    "kangaroo": "Kangaroo",
    "lion": "Lion",
    "manta-ray": "Manta Ray",
    "otter": "Otter",
    "pufferfish": "Pufferfish",
    "rockfish": "Rockfish",
    "salmon": "Salmon",
    "seahorse": "Seahorse",
    "triggerfish": "Triggerfish",
    "zebra": "Zebra",
  },
  "contestDescription": {
    "fishing": "Should fishing be banned in all city owned lakes and rivers?",
    "new-zoo-either": "Initiative Measure No. 12, Should Sample City establish a new safari-style zoo costing 2,000,000?

 Alternative Measure 12 A, Should Sample City establish a new traditional zoo costing 1,000,000?",
    "new-zoo-pick": "Initiative Measure No. 12, Should Sample City establish a new safari-style zoo costing 2,000,000?

 Alternative Measure 12 A, Should Sample City establish a new traditional zoo costing 1,000,000?",
  },
  "contestOptionLabel": {
    "allow-fishing": "NO",
    "ban-fishing": "YES",
    "new-zoo-either-approved": "FOR APPROVAL OF EITHER Initiative No. 12 OR Alternative Initiative No. 12 A",
    "new-zoo-neither-approved": "AGAINST BOTH Initiative No. 12 AND Alternative Measure 12 A",
    "new-zoo-safari": "FOR Initiative No. 12",
    "new-zoo-traditional": "FOR Alternative Measure No. 12 A",
  },
  "contestTerm": {
    "aquarium-council-fish": "For three years",
    "best-animal-fish": "For three years",
    "best-animal-mammal": "For three years",
    "zoo-council-mammal": "For three years",
  },
  "contestTitle": {
    "aquarium-council-fish": "Zoo Council",
    "best-animal-fish": "Best Animal",
    "best-animal-mammal": "Best Animal",
    "fishing": "Ballot Measure 3",
    "new-zoo-either": "Ballot Measure 1",
    "new-zoo-pick": "Ballot Measure 1",
    "zoo-council-mammal": "Zoo Council",
  },
  "countyName": "Sample County",
  "districtName": {
    "district-1": "District 1",
  },
  "electionDate": "September 8, 2021",
  "electionTitle": "Example Primary Election",
  "partyFullName": {
    "0": "Mammal Party",
    "1": "Fish Party",
  },
  "partyName": {
    "0": "Mammal",
    "1": "Fish",
  },
  "precinctName": {
    "precinct-1": "Precinct 1",
    "precinct-2": "Precinct 2",
  },
  "stateName": "State of Sample",
}
`;

exports[`extractAndTranslateElectionStrings should extract and translate election strings correctly for multiple languages 1`] = `
{
  "ballotLanguage": "espa√±ol (EE. UU.)",
  "contestDescription": {
    "water-1-fishing": "Should fishing be banned in all city owned lakes and rivers? (in es-US)",
    "water-2-fishing": "Should fishing be banned in all city owned lakes and rivers? (in es-US)",
  },
  "contestOptionLabel": {
    "water-1-fishing-allow-fishing": "NO (in es-US)",
    "water-1-fishing-ban-fishing": "YES (in es-US)",
    "water-2-fishing-allow-fishing": "NO (in es-US)",
    "water-2-fishing-ban-fishing": "YES (in es-US)",
  },
  "contestTitle": {
    "congressional-1-fish": "Congressional 1 Representative - Fish (in es-US)",
    "congressional-1-mammal": "Congressional 1 Representative - Mammal (in es-US)",
    "congressional-2-fish": "Congressional 2 Representative - Fish (in es-US)",
    "congressional-2-mammal": "Congressional 2 Representative - Mammal (in es-US)",
    "county-leader-fish": "County Leader Fish (in es-US)",
    "county-leader-mammal": "County Leader Mammal (in es-US)",
    "water-1-fishing": "Water District 1 Fishing (in es-US)",
    "water-2-fishing": "Water District 2 Fishing (in es-US)",
  },
  "countyName": "Sample County (in es-US)",
  "districtName": {
    "congressional-1": "Congressional District 1 (in es-US)",
    "congressional-2": "Congressional District 2 (in es-US)",
    "county": "County (in es-US)",
    "water-1": "Water District 1 (in es-US)",
    "water-2": "Water District 2 (in es-US)",
  },
  "electionDate": "8 de septiembre de 2021",
  "electionTitle": "Example Primary Election (in es-US)",
  "partyFullName": {
    "0": "Mammal Party (in es-US)",
    "1": "Fish Party (in es-US)",
  },
  "partyName": {
    "0": "Mammal (in es-US)",
    "1": "Fish (in es-US)",
  },
  "precinctName": {
    "precinct-c1-w1-1": "Precinct 1 (in es-US)",
    "precinct-c1-w1-2": "Precinct 2 (in es-US)",
    "precinct-c1-w2": "Precinct 3 (in es-US)",
    "precinct-c2": "Precinct 4 (in es-US)",
  },
  "stateName": "State of Sample (in es-US)",
}
`;
